### PR TITLE
MOTECH-1913: Import 'Tree View' tab from the whp-mTraining

### PR DESCRIPTION
**Pull request by tstalka**
Now we can use tree in the mTraining UI, saving is not supported yet

### CODE REVIEW

**Comment by tstalka** [comment]
test this please

**Comment by tstalka** [comment]
This pull request is now ready to review

File: mtraining/src/main/java/org/motechproject/mtraining/domain/Chapter.java [hunk 1]
```diff
@@ -81,4 +101,27 @@ public void setQuiz(Quiz quiz) {
...
+    }
+
+    public void setCourse(Course course) {
+        this.course = course;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Ignore
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/domain/Chapter.java]: not required

File: mtraining/src/main/java/org/motechproject/mtraining/domain/Course.java [hunk 2]
```diff
@@ -66,4 +74,19 @@ public void setChapters(List<Chapter> chapters) {
     public void setProperties(Map<String, String> properties) {
         this.properties = properties;
     }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Ignore
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/domain/Course.java]: not required

File: mtraining/src/main/java/org/motechproject/mtraining/domain/CourseUnitMetadata.java [hunk 3]
```diff
@@ -82,4 +86,37 @@ public String getDescription() {
     public void setDescription(String description) {
         this.description = description;
     }
+
+    /**
+     * Builds the dto representation of the unit used by the tree view.
+     *
+     * @return the dto representation of the unit
+     */
+    @Ignore
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/domain/CourseUnitMetadata.java]: not required

File: mtraining/src/main/java/org/motechproject/mtraining/domain/CourseUnitMetadata.java [hunk 4]
```diff
@@ -82,4 +86,37 @@ public String getDescription() {
...
+
+        return Objects.equals(this.getId(), other.getId())
+                && Objects.equals(this.getName(), other.getName())
+                && Objects.equals(this.getContent(), other.getContent())
+                && Objects.equals(this.getDescription(), other.getDescription());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getId());
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/domain/CourseUnitMetadata.java]: not consistent with equals, also the use of ID field should be avoided in the hashCode method

File: mtraining/src/main/java/org/motechproject/mtraining/domain/Quiz.java [hunk 5]
```diff
@@ -80,4 +93,45 @@ public void setPassPercentage(double passPercentage) {
...
+    }
+
+    public void setChapter(Chapter chapter) {
+        this.chapter = chapter;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Ignore
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/domain/Quiz.java]: not required

File: mtraining/src/main/java/org/motechproject/mtraining/domain/Quiz.java [hunk 6]
```diff
@@ -80,4 +93,45 @@ public void setPassPercentage(double passPercentage) {
...
+        return Objects.equals(this.getId(), other.getId())
+                && Objects.equals(this.getName(), other.getName())
+                && Objects.equals(this.getPassPercentage(), other.getPassPercentage())
+                && Objects.equals(this.getContent(), other.getContent())
+                && Objects.equals(this.getDescription(), other.getDescription());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getId());
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/domain/Quiz.java]: not consistent with equals, also the use of ID should be avoided

File: mtraining/src/main/java/org/motechproject/mtraining/dto/CourseUnitDeserializer.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+/**
+ * {@code CourseUnitDeserializer} for the {@code List<CourseUnitDto>}.
+ */
+public class CourseUnitDeserializer extends JsonDeserializer<List<CourseUnitDto>> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CourseUnitDeserializer.class);
+
+    private ObjectMapper mapper;
+    private List<CourseUnitDto> units;
+    private JsonNode jsonNode;
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/dto/CourseUnitDeserializer.java]: why are all of the above class fields, if we reinitialize them with every method call? Cant they be mehod local variables?

File: mtraining/src/main/java/org/motechproject/mtraining/exception/CourseUnitNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package org.motechproject.mtraining.exception;
+
+/**
+ * The <code>CourseUnitNotFoundException</code> exception signals a situation in which an course unit does not
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/exception/CourseUnitNotFoundException.java]: a course unit

File: mtraining/src/main/java/org/motechproject/mtraining/repository/query/UnusedUnitQueryExecution.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.motechproject.mds.query.QueryExecution;
+import org.motechproject.mds.util.InstanceSecurityRestriction;
+
+import javax.jdo.Query;
+import java.util.List;
+
+/**
+ * The <code>UnusedUnitQueryExecution</code> class prepares a custom MDS query. The query should return
+ * only these units, that are not used by higher-level units, for example Lessons which doesn't have relation with
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/repository/query/UnusedUnitQueryExecution.java]: lessons that don't have~

File: mtraining/src/main/java/org/motechproject/mtraining/service/impl/CourseStructureServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+            if (owner != null) {
+                owner.setQuiz(null);
+            }
+        }
+    }
+
+    private void checkMissing(Set<Long> expected, List<Long> actual, String type) {
+        for (Long id : expected) {
+            for (Long unitId : actual) {
+                if (id.equals(unitId)) {
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/service/impl/CourseStructureServiceImpl.java]: There's some flaw in logic here

File: mtraining/src/main/java/org/motechproject/mtraining/dto/CourseUnitListWrapper.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
+package org.motechproject.mtraining.dto;
+
+import java.util.List;
+
+/**
+ * Wrapper class for list of course units.
+ */
+public class CourseUnitListWrapper {
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/dto/CourseUnitListWrapper.java]: Can't this just extend List<CourseUnitDto>? Do we need additional field?

File: mtraining/src/main/resources/webapp/messages/messages.properties [hunk 12]
```diff
@@ -8,4 +8,26 @@ mtraining.chapters=Chapters
...
-mtraining.activityRecords=Activity Records - Logs
\ No newline at end of file
+mtraining.activityRecords=Activity Records - Logs
+
+#TreeView
+mtraining.active=Active
+mtraining.inactive=Inactive
+mtraining.pending=Pending
+mtraining.confirm=Confirm
+mtraining.confirm.removeMember=Do you want remove this member? Remember that all his children would be disconnected from it.
```
Comment (sebbrudzinski) [inline mtraining/src/main/resources/webapp/messages/messages.properties]: its children will be~

**Comment by sebbrudzinski** [comment]
Several improvements suggestions:
- The "change path state" and dropdown should be linked together somehow, to signal that they work together and that the change must be confirmed with the click of a button
- Dragging an dropping quiz on a chapter that already has a quiz should replace the quizes
- We should make it more obvious for the user, that they have to click chapters or courses in order to drag and drop members there

File: mtraining/src/main/java/org/motechproject/mtraining/service/impl/CourseStructureServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                    chapters.add(chapter);
+                    chapterMap.remove(chapterUnitDto.getId());
+                }
+            }
+
+            course.setState(CourseUnitState.valueOf(courseUnitDto.getState()));
+            course.setChapters(chapters);
+            courseMap.remove(courseUnitDto.getId());
+            // we must evict cache to avoid relations inconsistent
+            courseDataService.evictAllCache();
```
Comment (sebbrudzinski) [inline mtraining/src/main/java/org/motechproject/mtraining/service/impl/CourseStructureServiceImpl.java]: Didn't you mean to use evictEntityCache here and below? EvictAllCache clears the entire Datanucleus cache for all entities

**Comment by sebbrudzinski** [comment]
@tstalka Seems to work properly; I've left one comment about clearing cache

**Comment by pgesek** [comment]
test MOTECH-1913

**Comment by pgesek** [comment]
#MOTECH-1732 test

**Pull request merged**